server/src/main/java/io/crate/planner/operators/Union.java

```diff
@@ -204,11 +204,12 @@ public LogicalPlan replaceSources(List<LogicalPlan> sources) {
     @Override
     public LogicalPlan pruneOutputsExcept(Collection<Symbol> outputsToKeep) {
         IntArrayList outputIndicesToKeep = new IntArrayList();
+        final int[] idx = {0};
         for (Symbol outputToKeep : outputsToKeep) {
             SymbolVisitors.intersection(outputToKeep, outputs, s -> {
-                int idx = outputs.indexOf(s);
```

Comment:
When there are duplicates, `outputs.indexOf(s)` always finds the first occurrences, such that `outputIndicesToKeep` contains duplicate indices leading to pruning LHS and RHS.

The approach is to have a sort of watermark that any elements of `outputs` can only be accounted once. This logic works under assumption that `outputsToKeep` is a subset of `outputs` with the order preserved.

Reply:
`outputsToKeep` preserves the original order:
```
     * @param outputsToKeep The collection should provide those outputs in the same original order, to avoid errors
     *                      regarding {@link Union} (which requires the order to be kept), or introduction of
     *                      unnecessary {@link Eval} operators.
```

Reply:
Not taking this path.

Reply:
not taking this path